spec/zoneddatetime.html

```diff
@@ -1456,28 +1450,31 @@ <h1>
           1. Repeat, while _days_ &gt; 0 and ℝ(_relativeResult_.[[EpochNanoseconds]]) &gt; _endNs_,
             1. Set _days_ to _days_ - 1.
             1. Set _relativeResult_ to ? AddDaysToZonedDateTime(_startInstant_, _startDateTime_, _timeZoneRec_, _zonedRelativeTo_.[[Calendar]], _days_).
-        1. Set _nanoseconds_ to _endNs_ - ℝ(_relativeResult_.[[EpochNanoseconds]]).
+        1. Set _norm_ to NormalizedTimeDurationFromEpochNanosecondsDifference(_endNs_, _relativeResult_.[[EpochNanoseconds]]).
         1. Let _done_ be *false*.
         1. Let _dayLengthNs_ be ~unset~.
         1. Repeat, while _done_ is *false*,
           1. Let _oneDayFarther_ be ? AddDaysToZonedDateTime(_relativeResult_.[[Instant]], _relativeResult_.[[DateTime]], _timeZoneRec_, _zonedRelativeTo_.[[Calendar]], _sign_).
-          1. Set _dayLengthNs_ to ℝ(_oneDayFarther_.[[EpochNanoseconds]] - _relativeResult_.[[EpochNanoseconds]]).
-          1. If (_nanoseconds_ - _dayLengthNs_) &times; _sign_ &ge; 0, then
-            1. Set _nanoseconds_ to _nanoseconds_ - _dayLengthNs_.
+          1. Set _dayLengthNs_ to NormalizedTimeDurationFromEpochNanosecondsDifference(_oneDayFarther_.[[EpochNanoseconds]], _relativeResult_.[[EpochNanoseconds]]).
+          1. Let _oneDayLess_ be ? SubtractNormalizedTimeDuration(_norm_, _dayLengthNs_).
+          1. If NormalizedTimeDurationSign(_oneDayLess_) &times; _sign_ &ge; 0, then
+            1. Set _norm_ to _oneDayLess_.
             1. Set _relativeResult_ to _oneDayFarther_.
             1. Set _days_ to _days_ + _sign_.
           1. Else,
             1. Set _done_ to *true*.
         1. If _days_ &lt; 0 and _sign_ = 1, throw a *RangeError* exception.
         1. If _days_ &gt; 0 and _sign_ = -1, throw a *RangeError* exception.
-        1. If _nanoseconds_ &lt; 0, then
+        1. If NormalizedTimeDurationSign(_norm_) = -1, then
           1. Assert: _sign_ is -1.
-        1. If _nanoseconds_ &gt; 0 and _sign_ = -1, throw a *RangeError* exception.
-        1. Assert: The inequality abs(_nanoseconds_) &lt; abs(_dayLengthNs_) holds.
+        1. If NormalizedTimeDurationSign(_norm_) = 1 and _sign_ = -1, throw a *RangeError* exception.
+        1. Assert: CompareNormalizedTimeDuration(NormalizedTimeDurationAbs(_norm_), NormalizedTimeDurationAbs(_dayLengthNs_)) = -1.
+        1. Let _dayLength_ be abs(ℝ(_dayLengthNs_.[[TotalNanoseconds]])).
+        1. If _dayLength_ &ge; 2<sup>53</sup>, throw a *RangeError* exception.
         1. Return the Record {
           [[Days]]: _days_,
```

Comment:
Does it make sense to restrict `days` to the duration-days limit, i.e. throw an error if `days ≥ ⌈(2**53) / (24 * 60 * 60)⌉`?

The SpiderMonkey implementation will likely have this restriction, because it ensures addition can be computed using `int64_t` in RoundDuration/RoundNumberToIncrement.

Reply:
I'm tentatively assuming that it's not possible to get _days_ up to that large of a value in NormalizedTimeDurationToDays:

- In step 11, _days_ starts out as the difference between two PlainDates, so the extremum is ±2e8.
- In step 13, it can be decremented by 1.
- In step 14, it can be incremented by 1.
- In step 16, if it's positive, it can be reduced at most down to 0, but no farther.
- In the loop calculating the day length in step 20, it can be incremented arbitrarily in either direction, but it's limited long before it gets to that point because it will throw if abs(_dayLength_) becomes ≥ 2⁵³ ns.

So I'm assuming we can just add an assertion for clarity here — let me know if you disagree, you tend to be better at spotting these things than I am.

Reply:
> * In the loop calculating the day length in step 20, it can be incremented arbitrarily in either direction, but it's limited long before it gets to that point because it will throw if abs(_dayLength_) becomes ≥ 2⁵³ ns.

It used to be possible through user-defined calendars, because then it was possible to go into the wrong direction, i.e. `sign` was positive, but instead of getting a future date, the user-defined calendar returned a past date. This allowed to arbitrarily increase `days`. I guess this no longer possible since ec1df1f178578eecad0b027b8c96a20d30b2c590, when `AddDaysToZonedDateTime` was added, so we no longer call `AddZonedDateTime`.